text/0156-columnar-encoding.md

```diff
@@ -0,0 +1,1417 @@
+high variability of the nature of the attributes within the OTLP entities.
+
+```yaml
+# Attributes Arrow Schema (declaration used in other schemas)
+attributes: &attributes                                 # arrow type = map
+  - key: string_dictionary | string                     # string dictionary by default, fallback to a string column when the cardinality is too high
+    value:                                              # arrow type = sparse union
+      str: string_dictionary | string 
+      i64: int64
+      f64: float64
+      bool: bool 
+      binary: binary_dictionary | binary
+      cbor: binary                                      # cbor encoded complex attribute values
+```
+
+> Note: **Dense vs Sparse union**: Apache Arrow supports two types of union: dense and sparse. Dense unions are more efficient
+> in memory usage but are less efficient in terms of processing speed (and also in terms of compression ratio according
+> to some tests). The memory consumed by a sparse union depends directly on the number of variants present in the union
+> definition. It is therefore important to keep the number of variants low. The `cbor` variant groups all complex
+> attribute values that cannot be represented with the other available variants. This representation aims to minimize
+> the memory overhead related to the number of variants in the union. The current tradeoff optimizes the processing
+> speed and, to some extent, the compression ratio.
+> Optimizations on memory consumption are theoretically possible at the level of the Arrow Go library in order to
+> minimize the overhead of sparse unions in the future.
+
+#### Metrics Payload
+
+We start by defining the Arrow Schema of the `exemplar` concept because it is used for several types of metrics.
+
+```yaml
+# Exemplar Arrow Schema (declaration used in other schemas)
+exemplars: &exemplars             # arrow type = list of struct
+  - attributes: *attributes       # YAML alias to the attributes schema defined previously
+    time_unix_nano: timestamp     # arrow type = timestamp (time unit nanoseconds)
+    value:                        # arrow type = sparse union
+      i64: int64
+      f64: float64
+    span_id: 8_bytes_binary_dictionary | 8_bytes_binary       # arrow fixed size binary array
+    trace_id: 16_bytes_binary_dictionary | 16_bytes_binary    # arrow fixed size binary array
+```
+
+`span_id` and `trace_id` are represented as fixed size binary dictionaries by default but can evolve to non-dictionary
+form when their cardinality exceeds a certain threshold (usually 2^16).
+
+> Note: every OTLP timestamps are represented as Arrow timestamps with nanoseconds time unit. This representation will
+> simplify the integration with the rest of the Arrow ecosystem (numerous time/date functions are supported in
+> DataFusion for example).
+
+The Arrow Schema for the univariate metrics is the following:
+
+```yaml
```

Comment:
I am not familiar with Arrow. Are Arrow schemas typically specified in yaml format or yaml is chosen for readability of this spec?

Reply:
There is unfortunately no official Arrow schema file format. I used initially YAML files for readability but in the last version I used ER diagram to improve even further the readability.